### PR TITLE
Store user in storage

### PR DESCRIPTION
**Pull request by huynhnkhanh**
1. User go to home page -> redirect to login page 
- I created PrivateHomeRoute.js in the components folder
image
image
- Then I imported PrivateHomeRoute.js file inside App.js
image
2. User go to login page directly
- I created PrivateLoginRoute.js in the components folder
image
image
- Then I imported PrivateLoginRoute.js file inside App.js
image

### CODE REVIEW

File: src/components/PrivateHomeRoute.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+          getToken() ? <Component {...props} /> : <Redirect to="/login" />
+        )} />
+  )
+}
+
+PrivateHomeRoute.propTypes = {
+  component: PropTypes.func.isRequired
+}
+
+export default PrivateHomeRoute
```
Comment (thaoho21294) [inline src/components/PrivateHomeRoute.js]: Avoid too short components like this.
- For the case `unauthenticated`, you can validate it in the `App` component by returning the login page if no token.
- For the case `authenticated` and go to log in, you can check in the `Login` component by redirecting to `/home` if there's a token.

**Review by thaoho21294 — CHANGES_REQUESTED** [review]
For other minor mistakes, you should self-review your code again to fix them.

File: src/App.js [hunk 2]
```diff
@@ -1,19 +1,25 @@
...
-          <Route path='/login' component={Login } />
-          <Route exact path='/detail/:locationId/:year/:month/:day' component={WeatherHourlyList} />
+          <Route path='/login' component={Login} >
+            {getToken() ? <Redirect to='/' /> : <Login />}
+          </Route>
+          <Route exact path='/' component={Home}>
+            {getToken() ? <Home /> : <Redirect to='/login' />}
+            <Route path='/login' component={Login}></Route>
+          </Route>
+          <Route path='/detail/:locationId/:year/:month/:day' component={WeatherHourlyList} />
```
Comment (thaoho21294) [inline src/App.js]: Do you feel this is complicated?
Should implement this way.

- For the case unauthenticated, you can validate it in the App component by returning the login page if no token.
==> in App component, if no token, return login page
```js
function App () {
  if(!getToken()) {
    return <Login />
  }

  return (
    <div className='App'>
     ...
    </div>
  )
}
```

- For the case authenticated and go to log in, you can check in the Login component by redirecting to /home if there's a token.
==> in Login component, if getToken ok, redirect to Home
```js
const Login = () => {
  if(getToken) {
    window.location.replace('/')
  }
}
```
but this will be checked every time the Login component re-render when users type a username or password. It should check only the first render. Then you should find a solution.

**Review by thaoho21294 — CHANGES_REQUESTED** [review]

File: src/scenes/Login/Login.js [hunk 3]
```diff
@@ -1,12 +1,15 @@
 import React, { useState } from 'react'
 import './_login.scss'
 import PropTypes from 'prop-types'
+import { getToken, saveToken } from '../../useToken'
 
-const Login = ({ history }) => {
+const Login = () => {
+  if (getToken()) {
+    return window.location.replace('/')
+  }
```
Comment (thaoho21294) [inline src/scenes/Login/Login.js]: if you put the check here, `getToken()` will be called a lot of times. Especially when you typing and the form is re-rendered. You may want to verify it.
So you need to find a solution to avoid this performance issue. Think about when you need to check the token and how many times it should be called.

**Review by thaoho21294 — CHANGES_REQUESTED** [review]

File: src/App.js (hunk 2 above)
Comment (thaoho21294) [reply src/App.js]: > but this will be checked every time the Login component re-render when users type a username or password. It should check only the first render. Then you should find a solution.

Remember reading carefully ☝️

File: src/scenes/Login/Login.js [hunk 4]
```diff
@@ -31,8 +36,10 @@ const Login = ({ history }) => {
       username,
       password
     })
-    if (foundUser) {
-      return history.replace('/')
+    console.log(getToken())
```
Comment (thaoho21294) [inline src/scenes/Login/Login.js]: remove console.log

File: src/scenes/Login/Login.js [hunk 5]
```diff
@@ -1,18 +1,23 @@
...
     setUsername(e.target.value)
   }
   const handleChangePassword = e => {
     setPassword(e.target.value)
   }
+  useEffect(() => {
+    if (getToken()) {
+      return window.location.replace('/')
+    }
+  }, [])
```
Comment (thaoho21294) [inline src/scenes/Login/Login.js]: For a better structure, should move this section below the function `handleSubmit` and use a blank line to separate them
Good structure should be:
- Variables declaration section
- Function declaration section
- hook section

File: src/scenes/Login/Login.js [hunk 6]
```diff
@@ -31,8 +36,10 @@ const Login = ({ history }) => {
       username,
       password
     })
-    if (foundUser) {
-      return history.replace('/')
+    console.log(getToken())
+    saveToken(foundUser)
+    if (getToken()) {
+      return window.location.replace('/')
```
Comment (thaoho21294) [inline src/scenes/Login/Login.js]: Why did you remove the old logic? Just check the found user, don't need to get the token then check the token. This takes more steps. Should be 👇 
submit form -> find user -> if(foundUser) -> saveToken and replace('/')

File: src/scenes/Login/Login.js [hunk 7]
```diff
@@ -56,15 +63,15 @@ const Login = ({ history }) => {
...
+          Forgot <a href='#'>password?</a>
         </p>
       </form>
     </div>
   )
 }
 
 Login.propTypes = {
-  history: PropTypes.object.isRequired
+  setToken: PropTypes.func.isRequired
```
Comment (thaoho21294) [inline src/scenes/Login/Login.js]: As I see Login doesn't have any props, huh?

File: src/useToken.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  const tokenString = localStorage.getItem('token')
+  return JSON.parse(tokenString)
+}
+const saveToken = userToken => {
+  localStorage.setItem('token', JSON.stringify(userToken))
+}
+export {
+  getToken,
+  saveToken
+}
```
Comment (thaoho21294) [inline src/useToken.js]: should add a blank line between functions and the export section.

File: src/App.js [hunk 9]
```diff
@@ -5,15 +5,19 @@ import './App.css'
 import Home from './scenes/Home/Home'
 import WeatherHourlyList from './scenes/WeatherHourlyList/WeatherHourlyList'
 import Login from './scenes/Login/Login'
+import { getToken } from './useToken'
```
Comment (thaoho21294) [inline src/App.js]: rename the file to `auth.js`. This is not a custom hook so should not have the prefix `use`

**Review by thaoho21294 — CHANGES_REQUESTED** [review]

File: src/scenes/Login/Login.js (hunk 5 above)
Comment (thaoho21294) [reply src/scenes/Login/Login.js]: And no need to `return`

File: src/scenes/Login/Login.js (hunk 6 above)
Comment (thaoho21294) [reply src/scenes/Login/Login.js]: Same at here, no need to `return`

File: src/scenes/Login/Login.js (hunk 4 above)
Comment (huynhnkhanh) [reply src/scenes/Login/Login.js]: image

File: src/scenes/Login/Login.js (hunk 5 above)
Comment (huynhnkhanh) [reply src/scenes/Login/Login.js]: image

File: src/scenes/Login/Login.js (hunk 6 above)
Comment (huynhnkhanh) [reply src/scenes/Login/Login.js]: image

File: src/scenes/Login/Login.js (hunk 7 above)
Comment (huynhnkhanh) [reply src/scenes/Login/Login.js]: image
image

File: src/useToken.js (hunk 8 above)
Comment (huynhnkhanh) [reply src/useToken.js]: image

File: src/App.js (hunk 9 above)
Comment (huynhnkhanh) [reply src/App.js]: image

File: src/scenes/Login/Login.js [hunk 10]
```diff
@@ -32,9 +31,15 @@ const Login = ({ history }) => {
       password
     })
     if (foundUser) {
-      return history.replace('/')
+      saveToken(foundUser)
+      return window.location.replace('/')
```
Comment (thaoho21294) [inline src/scenes/Login/Login.js]: no need return

File: src/scenes/Login/Login.js [hunk 11]
```diff
@@ -32,9 +31,15 @@ const Login = ({ history }) => {
...
     }
   }
+  useEffect(() => {
+    if (getToken()) {
+      return window.location.replace('/')
```
Comment (thaoho21294) [inline src/scenes/Login/Login.js]: same at here

**Review by thaoho21294 — CHANGES_REQUESTED** [review]

**Review by thaoho21294 — APPROVED** [review]
LGTM, just missing some blank line
